File: src-tauri/src/commands/question.rs
use tauri::State;
use std::collections::HashMap;
use crate::state::AppState;
use crate::error::{AppError, AppResult};
use crate::ai::llm_client::{Message, LLMOptions};
use crate::ai::prompts;

/// 生成练习题
///
/// 根据学生画像清晰度自动选择模式：
/// - 冷启动（答题 < 5）→ 诊断模式（每单元 1 题）
/// - 初步画像（5~20）→ 按单元出题
/// - 画像清晰（> 20）→ 自适应出题
#[tauri::command]
pub async fn generate_quiz(
    student_id: String,
    knowledge_ids: Vec<String>,
    count: Option<i32>,
    unit: Option<String>,
    state: State<'_, AppState>,
) -> AppResult<serde_json::Value> {
    let count = count.unwrap_or(5) as usize;
    tracing::debug!("为学生 {} 生成 {} 道题, unit={:?}", student_id, count, unit);

    // 查询学生答题数据量以判断画像清晰度
    let (total_answers, mastery_data) = {
        let db = state.db.lock().map_err(|e| AppError::Internal(e.to_string()))?;

        // 统计答题总数
        let total: i64 = db
            .query_row(
                "SELECT COUNT(*) FROM answer_records WHERE student_id = ?1",
                rusqlite::params![student_id],
                |row| row.get(0),
            )
            .unwrap_or(0);

        // 获取各单元掌握度
        let mut mastery: HashMap<String, f64> = HashMap::new();
        if let Ok(mut stmt) = db.prepare(
            "SELECT kn.name, km.mastery_score
             FROM knowledge_mastery km
             JOIN knowledge_nodes kn ON km.knowledge_id = kn.id
             WHERE km.student_id = ?1"
        ) {
            if let Ok(rows) = stmt.query_map(rusqlite::params![student_id], |row| {
                Ok((row.get::<_, String>(0)?, row.get::<_, f64>(1)?))
            }) {
                for row in rows.flatten() {
                    mastery.insert(row.0, row.1);
                }
            }
        }

        (total, mastery)
    };

    let qb = &state.question_bank;

    // 确定出题模式
    let (mode, questions) = if let Some(ref unit_name) = unit {
        // 指定单元出题
        let qs = qb.quiz_for_unit(unit_name, count);
        ("unit", qs)
    } else if total_answers < 5 {
        // 冷启动诊断模式
        let qs = qb.diagnose_quiz(1);
        ("diagnose", qs)
    } else if total_answers < 20 || mastery_data.is_empty() {
        // 初步画像 — 按单元顺序
        let qs = qb.diagnose_quiz(2);
        ("emerging", qs)
    } else {
        // 画像清晰 — 自适应
        let qs = qb.adaptive_quiz(&mastery_data, count);
        ("adaptive", qs)
    };

    tracing::info!(
        "学生 {} 出题模式: {}, 生成 {} 道题",
        student_id,
        mode,
        questions.len()
    );

    let questions_json: Vec<serde_json::Value> = questions
        .iter()
        .map(|q| {
            serde_json::json!({
                "id": q.id,
                "unit": q.unit,
                "semester": q.semester,
                "question_type": q.question_type,
                "content_latex": q.content_latex,
                "answer_latex": q.answer_latex,
                "difficulty": q.difficulty,
            })
        })
        .collect();

    Ok(serde_json::json!({
        "mode": mode,
        "total": questions_json.len(),
        "questions": questions_json,
    }))
}

/// AI 动态出题
///
/// 行为：
/// 1. 取学生薄弱知识点（mastery_score 最低的 3 个）作为提示
/// 2. 调 LLM 生成一道符合 unit + difficulty 的题
/// 3. 解析 JSON，给题分配一个 ai-xxxx 的 id，注入到 question_bank（内存）以便后续判题
///    （注：因 question_bank 是只读 Arc，这里返回题目数据让前端塞进 store；
///     后端再用一个 ai_questions DashMap 缓存题目，submit_answer 才能找到它）
#[tauri::command]
pub async fn generate_ai_question(
    student_id: String,
    unit: String,
    difficulty: Option<i32>,
    state: State<'_, AppState>,
) -> AppResult<serde_json::Value> {
    let difficulty = difficulty.unwrap_or(2).clamp(1, 5);
    tracing::info!("AI 出题: student={} unit={} difficulty={}", student_id, unit, difficulty);

    // 取学生年级 + 薄弱知识点
    let (grade, weak_topics) = {
        let db = state.db.lock().map_err(|e| AppError::Internal(e.to_string()))?;
        let grade: i32 = db.query_row(
            "SELECT grade FROM students WHERE id = ?1",
            rusqlite::params![student_id],
            |row| row.get(0),
        ).unwrap_or(6);

        let mut topics: Vec<String> = Vec::new();
        if let Ok(mut stmt) = db.prepare(
            "SELECT COALESCE(kn.name, km.knowledge_id) FROM knowledge_mastery km
             LEFT JOIN knowledge_nodes kn ON km.knowledge_id = kn.id
             WHERE km.student_id = ?1 ORDER BY km.mastery_score ASC LIMIT 3"
        ) {
            if let Ok(rows) = stmt.query_map(rusqlite::params![student_id], |row| row.get::<_, String>(0)) {
                for r in rows.flatten() { topics.push(r); }
            }
        }
        (grade, topics)
    };

    // LLM 出题
    let llm_clone = {
        let guard = state.llm_client.lock().map_err(|e: std::sync::PoisonError<_>| AppError::Internal(e.to_string()))?;
        guard.clone()
    };
    let llm = match llm_clone {
        Some(c) => c,
        None => return Err(AppError::LLMError("AI 出题需要先在设置页配置 API Key".to_string())),
    };

    let prompt = prompts::generate_question(grade, &unit, difficulty, &weak_topics);
    let messages = vec![
        Message {
            role: "system".to_string(),
            content: "你是数学命题老师，必须以严格 JSON 输出，不要任何其他文字。".to_string(),
        },
        Message { role: "user".to_string(), content: prompt },
    ];
    let opts = LLMOptions {
        temperature: Some(0.8),
        max_tokens: Some(500),
        top_p: Some(0.9),
    };

    let raw = llm.complete(&messages, &opts).await
        .map_err(|e| AppError::LLMError(format!("LLM 出题失败: {}", e)))?;

    let parsed = parse_ai_question_json(&raw)
        .ok_or_else(|| AppError::LLMError(format!("LLM 输出无法解析为题目 JSON: {}", &raw[..raw.len().min(200)])))?;

    // 生成 id 并写入 ai_questions 缓存（让 submit_answer 能找到）
    let q_id = format!("ai-{}", uuid::Uuid::new_v4().to_string().split('-').next().unwrap_or("xxxx"));
    let q_type = parsed.get("question_type").and_then(|v| v.as_str()).unwrap_or("填空题").to_string();
    let content = parsed.get("content_latex").and_then(|v| v.as_str()).unwrap_or("").to_string();
    let answer = parsed.get("answer_latex").and_then(|v| v.as_str()).unwrap_or("").to_string();
    let unit_out = parsed.get("unit").and_then(|v| v.as_str()).unwrap_or(&unit).to_string();
    let hint = parsed.get("hint").and_then(|v| v.as_str()).unwrap_or("").to_string();

    let ai_q = crate::services::question_bank::BaseQuestion {
        id: q_id.clone(),
        unit: unit_out.clone(),
        semester: "AI".to_string(),
        question_type: q_type.clone(),
        content_latex: content.clone(),
        answer_latex: answer.clone(),
        difficulty,
    };
    state.ai_questions.insert(q_id.clone(), ai_q);

    Ok(serde_json::json!({
        "id": q_id,
        "unit": unit_out,
        "semester": "AI",
        "question_type": q_type,
        "content_latex": content,
        "answer_latex": answer,
        "difficulty": difficulty,
        "hint": hint,
        "ai_generated": true,
    }))
}

/// 解析 LLM 出题返回的 JSON（兼容裸 JSON / ```json 包裹 / 花括号片段）
fn parse_ai_question_json(raw: &str) -> Option<serde_json::Value> {
    let trimmed = raw.trim();
    if let Ok(v) = serde_json::from_str::<serde_json::Value>(trimmed) {
        if v.is_object() && v.get("content_latex").is_some() { return Some(v); }
    }
    if let Some(start) = trimmed.find("```json") {
        let after = &trimmed[start + 7..];
        if let Some(end) = after.find("```") {
            if let Ok(v) = serde_json::from_str::<serde_json::Value>(after[..end].trim()) {
                if v.is_object() { return Some(v); }
            }
        }
    }
    if let (Some(s), Some(e)) = (trimmed.find('{'), trimmed.rfind('}')) {
        if e > s {
            if let Ok(v) = serde_json::from_str::<serde_json::Value>(&trimmed[s..=e]) {
                if v.is_object() { return Some(v); }
            }
        }
    }
    None
}

/// 获取题库概览
#[tauri::command]
pub async fn get_question_bank_overview(
    state: State<'_, AppState>,
) -> AppResult<serde_json::Value> {
    let qb = &state.question_bank;
    let units: Vec<serde_json::Value> = qb
        .unit_names()
        .iter()
        .map(|name| {
            let qs = qb.questions_for_unit(name);
            serde_json::json!({
                "unit": name,
                "count": qs.len(),
                "types": qs.iter().map(|q| q.question_type.as_str()).collect::<Vec<_>>(),
            })
        })
        .collect();

    Ok(serde_json::json!({
        "total_questions": qb.questions.len(),
        "total_units": units.len(),
        "units": units,
    }))
}

File: server/app/routes/question.py
"""出题路由 - 静态题库 + LLM 动态出题"""
import json
import logging
import uuid
from functools import lru_cache

import aiosqlite
from fastapi import APIRouter, Depends, HTTPException

from ..ai.llm_client import LLMOptions, Message, get_llm_client
from ..ai.prompts import generate_question
from ..config import Settings, get_settings
from ..db.connection import get_db
from ..schemas.common import ok
from ..schemas.question import (
    AIQuestionRequest,
    QuestionContent,
    QuizGenerateRequest,
)
from ..services.question_bank import QuestionBank
from ._deps import require_api_key

logger = logging.getLogger(__name__)
router = APIRouter(dependencies=[Depends(require_api_key)])


@lru_cache(maxsize=2)
def _load_bank(path: str) -> QuestionBank:
    return QuestionBank.from_file(path)


def _bank_from_settings(settings: Settings) -> QuestionBank:
    return _load_bank(str(settings.question_bank_file))


def _q_to_schema(q) -> dict:
    return {
        "id": q.id,
        "unit": q.unit,
        "semester": q.semester,
        "question_type": q.question_type,
        "content_latex": q.content_latex,
        "answer_latex": q.answer_latex,
        "difficulty": q.difficulty,
        "knowledge_point": getattr(q, "knowledge_point", None),
    }


@router.get("/questions/overview")
async def bank_overview(settings: Settings = Depends(get_settings)):
    try:
        bank = _bank_from_settings(settings)
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"题库加载失败: {e}") from e
    units = bank.unit_names()
    return ok(
        {
            "total": len(bank.questions),
            "unit_count": len(units),
            "units": units,
        }
    )


@router.post("/questions/quiz")
async def generate_quiz(
    payload: QuizGenerateRequest,
    db: aiosqlite.Connection = Depends(get_db),
    settings: Settings = Depends(get_settings),
):
    """生成练习题

    对齐 Rust 版 `commands/question.rs::generate_quiz` 的行为：

    - 客户端显式指定 mode='unit' 且带 unit 名 → 按单元出题
    - 否则按学生答题总数自动切换：
        - 答题 < 5     → diagnose（每单元抽 1）
        - 答题 5~19    → emerging（每单元抽 2）
        - 答题 ≥ 20    → adaptive（按掌握度加权）
    - 若客户端显式指定 mode='diagnose' / 'adaptive' 也会被尊重

    mastery_data 的 key 用 `knowledge_nodes.name`（和 Rust 版保持一致），
    question_bank.adaptive_quiz 内部用题目的 `unit` 字段查找此 dict。
    """
    bank = _bank_from_settings(settings)

    # 1. 统计答题总数（决定自动模式）
    async with db.execute(
        "SELECT COUNT(*) AS cnt FROM answer_records WHERE student_id = ?",
        (payload.student_id,),
    ) as cur:
        row = await cur.fetchone()
    total_answers = int(row["cnt"] or 0) if row else 0

    # 2. 读掌握度（用 knowledge_node.name 作为 key）
    async with db.execute(
        """
        SELECT kn.name AS knowledge_name, km.mastery_score
        FROM knowledge_mastery km
        JOIN knowledge_nodes kn ON kn.id = km.knowledge_id
        WHERE km.student_id = ?
        """,
        (payload.student_id,),
    ) as cur:
        mastery_rows = await cur.fetchall()
    mastery_data: dict[str, float] = {
        r["knowledge_name"]: float(r["mastery_score"]) for r in mastery_rows
    }

    # 3. 确定出题模式
    requested = payload.mode or "auto"
    if requested == "unit" and payload.unit:
        mode, qs = "unit", bank.quiz_for_unit(payload.unit, payload.count)
    elif requested == "diagnose":
        mode, qs = "diagnose", bank.diagnose_quiz(count_per_unit=1)
    elif requested == "adaptive":
        mode, qs = "adaptive", bank.adaptive_quiz(mastery_data, payload.count)
    else:
        # auto: 按学生画像清晰度切换
        if total_answers < 5:
            mode, qs = "diagnose", bank.diagnose_quiz(count_per_unit=1)
        elif total_answers < 20 or not mastery_data:
            mode, qs = "emerging", bank.diagnose_quiz(count_per_unit=2)
        else:
            mode, qs = "adaptive", bank.adaptive_quiz(mastery_data, payload.count)

    logger.info(
        f"student={payload.student_id} total_answers={total_answers} "
        f"mode={mode} questions={len(qs)}"
    )
    return ok(
        {"mode": mode, "questions": [_q_to_schema(q) for q in qs]}
    )


@router.post("/questions/ai")
async def generate_ai_question(
    payload: AIQuestionRequest,
    db: aiosqlite.Connection = Depends(get_db),
):
    """调 LLM 动态生成一道题，并立刻落库到 questions 表

    落库原因：后续 submit_answer 需要引用 question_id，而 answer_records.question_id
    有 FK 约束指向 questions(id)。如果不落库，答题时会触发 FOREIGN KEY constraint
    failed（Bug #1）。
    """
    client = get_llm_client()
    if not client.api_key:
        raise HTTPException(status_code=500, detail="未配置 SILICONFLOW_API_KEY")

    prompt = generate_question(
        grade=6,  # 当前只支持 6 年级
        unit=payload.unit,
        difficulty=payload.difficulty,
        weak_topics=payload.weak_topics,
        interest_context="",
    )

    text = await client.complete(
        messages=[Message(role="user", content=prompt)],
        opts=LLMOptions(temperature=0.9, max_tokens=1024),
    )

    # 尝试提取 JSON
    cleaned = text.strip()
    if cleaned.startswith("```"):
        # 去掉可能的 ```json 包裹
        parts = cleaned.split("```")
        if len(parts) >= 3:
            cleaned = parts[1]
            if cleaned.startswith("json"):
                cleaned = cleaned[4:]
            cleaned = cleaned.strip()

    try:
        data = json.loads(cleaned)
    except json.JSONDecodeError as e:
        logger.error(f"LLM 出题响应解析失败: {text}")
        raise HTTPException(
            status_code=502, detail=f"LLM 响应不是合法 JSON: {e}"
        ) from e

    # ---- 落库：先 upsert knowledge_node，再 insert question ----
    # 用 "ai-<uuid>" 作为 id，和 Rust 版保持一致（submit_answer 里会识别 "ai-" 前缀）
    question_id = f"ai-{uuid.uuid4().hex[:12]}"
    unit_name = str(data.get("unit", payload.unit))
    knowledge_id = "kn-" + unit_name.replace(" ", "_")

    await db.execute(
        """
        INSERT OR IGNORE INTO knowledge_nodes
          (id, name, grade, semester, unit, sort_order)
        VALUES (?, ?, 6, 1, 0, 0)
        """,
        (knowledge_id, unit_name),
    )
    await db.execute(
        """
        INSERT INTO questions
          (id, knowledge_id, question_type, difficulty, content, answer, source)
        VALUES (?, ?, ?, ?, ?, ?, 'ai')
        """,
        (
            question_id,
            knowledge_id,
            str(data.get("question_type", "")),
            int(data.get("difficulty", payload.difficulty)),
            str(data.get("content_latex", "")),
            str(data.get("answer_latex", "")),
        ),
    )
    await db.commit()

    # 把生成的 id 塞回响应给客户端
    data["id"] = question_id
    data["knowledge_id"] = knowledge_id
    data["ai_generated"] = True
    return ok(data)
